tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,441 @@
+    } else if (rhs.IsPerTensorQuantized()) {
+      if (!output.IsPerTensorQuantized()) {
+        return absl::FailedPreconditionError(
+            "stablehlo.dot_general: If lhs and rhs are per-tensor quantized "
+            "than output should also be per-tensor quantized.");
+      }
+      if (lhs.quantized_per_tensor_element_type().ExpressedType() ==
+          rhs.quantized_per_tensor_element_type().ExpressedType()) {
+        if (lhs.quantized_per_tensor_element_type().ExpressedType() !=
+            output.quantized_per_tensor_element_type().ExpressedType()) {
+          return absl::FailedPreconditionError(
+              "stablehlo.dot_general: The expressed_type of output tensor must "
+              "be the same as the expressed_type of lhs and rhs tensors.");
+        }
+      }
+      auto check_zero_point_value_is_zero =
+          [](auto zero_point) -> bool {
+        if (std::holds_alternative<I4>(zero_point)) {
+          return std::get<I4>(zero_point) == static_cast<I4>(0);
+        } else if (std::holds_alternative<int8_t>(zero_point)) {
+          return std::get<int8_t>(zero_point) == static_cast<int8_t>(0);
+        } else if (std::holds_alternative<int16_t>(zero_point)) {
+          return std::get<int16_t>(zero_point) == static_cast<int16_t>(0);
+        }
+        return false;
+      };
+      if (!check_zero_point_value_is_zero(
+              rhs.quantized_per_tensor_element_type().ZeroPoint())) {
+        return absl::FailedPreconditionError(
+            "stablehlo.dot_general: The rhs per-tensor should have zero points "
+            "as 0.");
+      }
+    } else if (rhs.IsPerAxisQuantized()) {
+      if (output.IsPerTensorQuantized()) {
+        if (lhs.quantized_per_tensor_element_type().ExpressedType() ==
+            rhs.quantized_per_axis_element_type().ExpressedType()) {
+          if (lhs.quantized_per_tensor_element_type().ExpressedType() !=
+              output.quantized_per_tensor_element_type().ExpressedType()) {
+            return absl::FailedPreconditionError(
+                "stablehlo.dot_general: The expressed_type of output must be "
+                "the same as the expressed_type of lhs and rhs.");
+          }
+        }
+      } else if (output.IsPerAxisQuantized()) {
+        if (lhs.quantized_per_tensor_element_type().ExpressedType() ==
+            rhs.quantized_per_axis_element_type().ExpressedType()) {
+          if (lhs.quantized_per_tensor_element_type().ExpressedType() !=
+              output.quantized_per_axis_element_type().ExpressedType()) {
+            return absl::FailedPreconditionError(
+                "stablehlo.dot_general: The expressed_type of output must be "
+                "the same as the expressed_type of lhs and rhs.");
+          }
+        }
+      }
+      auto check_zero_points_values_are_zero =
+          [](auto zero_points) -> bool {
+        if (std::holds_alternative<absl::InlinedVector<I4, 8>>(zero_points)) {
+          auto zero_point_value =
```

Comment:
`auto&`, otherwise this will make a copy of the vector.

Also add `const` when you don't plan to modify the value.